### PR TITLE
Feat/discounts and refunds crud

### PR DESCRIPTION
**Pull request by aiste-i**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -401,3 +401,4 @@ FodyWeavers.xsd
 
 # JetBrains Rider
 *.sln.iml
+Looms.PoS/Looms.PoS/docker-compose.yml
```
Comment (titasstank) [inline .gitignore:404]: why add this?

File: Looms.PoS/Looms.PoS.Persistance/Repositories/RefundsRepository.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public async Task<RefundDao> CreateAsync(RefundDao refundDao)
+    {
+        var entityEntry = await _context.AddAsync(refundDao);
+        await _context.SaveChangesAsync();
+        return entityEntry.Entity;
+    }
+
+    // public async Task<RefundDao> CreateForPreviousPaymentAsync(RefundDao refundDao)
```
Comment (titasstank) [inline Looms.PoS/Looms.PoS.Persistance/Repositories/RefundsRepository.cs:23]: Remove comments

File: Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/CreateDiscount/CreateDiscountsCommand.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using Looms.PoS.Application.Abstracts;
+using MediatR;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Looms.PoS.Application.Features.Discount.Commands.CreateDiscount;
```
[the remark this answers is not in the document]
Comment (titasstank) [reply Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/CreateDiscount/CreateDiscountsCommand.cs:6]: Apply this fix to all files where applicable

File: .gitignore (hunk 1 above)
Comment (aiste-i) [reply .gitignore]: removed

File: Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/CreateDiscount/CreateDiscountsCommand.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/CreateDiscount/CreateDiscountsCommand.cs:6]: seems to be added everywhere

File: Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/CreateDiscount/CreateDiscountsCommandHandler.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+using MediatR;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Looms.PoS.Application.Features.Discount.Commands.CreateDiscount;
+public class CreateDiscountsCommandHandler : IRequestHandler<CreateDiscountsCommand, IActionResult>
+{
+    private readonly IDiscountsRepository _discountsRepository;
+    private readonly IHttpContentResolver _httpContentResolver;
+    private readonly IDiscountModelsResolver _modelsResolver;
+
```
[the remark this answers is not in the document]
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/CreateDiscount/CreateDiscountsCommandHandler.cs:15]: fixed in fixes from PR

File: Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/UpdateDiscount/UpdateDiscountCommandHandler.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        _modelsResolver = modelsResolver;
+    }
+
+    public async Task<IActionResult> Handle(UpdateDiscountCommand command, CancellationToken cancellationToken)
+    {
+        var discountRequest = await _httpContentResolver.GetPayloadAsync<UpdateDiscountRequest>(command.Request);
+        
+        var discountDao = _modelsResolver.GetDaoFromRequest(discountRequest);
+        Console.WriteLine(discountDao.Id);
+        Console.WriteLine(command.Id);
```
[the remark this answers is not in the document]
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/UpdateDiscount/UpdateDiscountCommandHandler.cs]: fixed in fixes from PR

File: Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/UpdateDiscount/UpdateDiscountCommandHandler.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        if(discountDao.Id.ToString() != command.Id)
```
[the remark this answers is not in the document]
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/UpdateDiscount/UpdateDiscountCommandHandler.cs]: removed in fixes from PR

File: Looms.PoS/Looms.PoS.Application/Features/Refund/Commands/CreateRefund/CreateRefundRequestValidator.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        RuleFor(x => x.Amount)
+            .GreaterThan(0);
+
+        RuleFor(x => x.RefundReason)
+            .NotEmpty();
+
+        RuleFor(x => x.OrderId)
+            .NotEmpty()
+            .Must(x => Guid.TryParse(x, out _) && x.Length == 36);
```
[the remark this answers is not in the document]
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Features/Refund/Commands/CreateRefund/CreateRefundRequestValidator.cs]: fixed in fixes from PR

File: Looms.PoS/Looms.PoS.Application/Features/Refund/Queries/GetRefund/GetRefundQueryValidator.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
+using FluentValidation;
+using Looms.PoS.Application.Features.Refund.Queries.GetRefund;
+using Looms.PoS.Application.Utilities.Validators;
+
+namespace Looms.PoS.Application.Features.Refund.Queries.GetRefunds;
+
+public class GetRefundsQueryValidator : AbstractValidator<GetRefundQuery>
```
[the remark this answers is not in the document]
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Features/Refund/Queries/GetRefund/GetRefundQueryValidator.cs]: fixed in fixes from PR

File: Looms.PoS/Looms.PoS.Application/Models/Requests/CreateDiscountRequest.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+namespace Looms.PoS.Application.Models.Requests;
+public record CreateDiscountRequest
+{
+    public string? Name { get; init; } = string.Empty;
+    public DiscountType DiscountType { get; init; }
+    public decimal Value { get; init; }
+    public DiscountTarget DiscountTarget { get; init; }
+    public Guid? ProductId { get; init; }
+    public DateTime StartDate { get; init; } 
+    public DateTime EndDate { get; init; }
```
[the remark this answers is not in the document]
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Models/Requests/CreateDiscountRequest.cs]: only enums are left as they were

File: Looms.PoS/Looms.PoS.Domain/Interfaces/IRefundsRepository.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+﻿using Looms.PoS.Domain.Daos;
+
+namespace Looms.PoS.Domain.Interfaces;
+public interface IRefundsRepository
+{
+    Task<RefundDao> CreateAsync(RefundDao refundDao);
+    
+    // Task<RefundDao> CreateForPreviousPaymentAsync(RefundDao refundDao);
```
[the remark this answers is not in the document]
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Domain/Interfaces/IRefundsRepository.cs]: fixed in fixes from PR

File: Looms.PoS/Looms.PoS.Persistance/Repositories/RefundsRepository.cs (hunk 2 above)
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Persistance/Repositories/RefundsRepository.cs]: fixed in fixes from PR

File: Looms.PoS/Looms.PoS.Application/Features/Refund/Queries/GetRefunds/GetRefundsQuery.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+using MediatR;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Looms.PoS.Application.Features.Refund.Queries.GetRefunds;
+
+public record GetRefundsQuery : IRequest<IActionResult>
```
Comment (titasstank) [inline Looms.PoS/Looms.PoS.Application/Features/Refund/Queries/GetRefunds/GetRefundsQuery.cs:6]: Inherit LoomHttpRequest, will be needed later with permissions

File: Looms.PoS/Looms.PoS.Application/Features/Refund/Queries/GetRefunds/GetRefundsQueryHandler.cs [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+using MediatR;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Looms.PoS.Application.Features.Refund.Queries.GetRefunds;
+
+public class GetRefundsQueryHandler : IRequestHandler<GetRefundsQuery, IActionResult>
+{
+    private readonly IRefundsRepository _refundsRepository;
+    private readonly IRefundModelsResolver _modelsResolver;
+    public GetRefundsQueryHandler(IRefundsRepository refundsRepository, IRefundModelsResolver modelsResolver)
```
Comment (titasstank) [inline Looms.PoS/Looms.PoS.Application/Features/Refund/Queries/GetRefunds/GetRefundsQueryHandler.cs:12]: empty line after properties

File: Looms.PoS/Looms.PoS.Application/Models/Requests/CreateDiscountRequest.cs [hunk 13]
```diff
@@ -0,0 +1,14 @@
+﻿using Looms.PoS.Domain.Enums;
+
+namespace Looms.PoS.Application.Models.Requests;
+
+public record CreateDiscountRequest
+{
+    public string? Name { get; init; } = string.Empty;
+    public DiscountType DiscountType { get; init; }
```
Comment (titasstank) [inline Looms.PoS/Looms.PoS.Application/Models/Requests/CreateDiscountRequest.cs:8]: Change enum types to string and validate in validator

**Review by titasstank — CHANGES_REQUESTED** [review]

File: Looms.PoS/Looms.PoS.Application/Models/Requests/CreateDiscountRequest.cs (hunk 13 above)
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Models/Requests/CreateDiscountRequest.cs]: changed to string for both request and response

File: Looms.PoS/Looms.PoS.Application/Features/Refund/Queries/GetRefunds/GetRefundsQueryHandler.cs (hunk 12 above)
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Features/Refund/Queries/GetRefunds/GetRefundsQueryHandler.cs:13]: added empty line

File: Looms.PoS/Looms.PoS.Application/Features/Refund/Queries/GetRefunds/GetRefundsQuery.cs (hunk 11 above)
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Features/Refund/Queries/GetRefunds/GetRefundsQuery.cs]: added LoomHttpRequest

File: Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/UpdateDiscount/UpdateDiscountCommandHandler.cs [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+using Microsoft.AspNetCore.Mvc;
+
+namespace Looms.PoS.Application.Features.Discount.Commands.UpdateDiscount;
+
+public class UpdateDiscountCommandHandler : IRequestHandler<UpdateDiscountCommand, IActionResult>
+{
+    private readonly IDiscountsRepository _discountsRepository;
+    private readonly IHttpContentResolver _httpContentResolver;
+    private readonly IDiscountModelsResolver _modelsResolver;
+
```
[the remark this answers is not in the document]
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Features/Discount/Commands/UpdateDiscount/UpdateDiscountCommandHandler.cs:15]: removed

File: Looms.PoS/Looms.PoS.Application/Models/Requests/CreateDiscountRequest.cs (hunk 13 above)
Comment (titasstank) [reply Looms.PoS/Looms.PoS.Application/Models/Requests/CreateDiscountRequest.cs]: response should be the correct type but make sure it outputs the string value and not a number

Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Models/Requests/CreateDiscountRequest.cs]: changed back to enums in response

File: Looms.PoS/Looms.PoS.Application/Features/Discount/Queries/GetDiscounts/GetDiscountsQuery.cs [hunk 15]
```diff
@@ -0,0 +1,12 @@
+﻿using Looms.PoS.Application.Abstracts;
+using MediatR;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Looms.PoS.Application.Features.Discount.Queries;
```
Comment (titasstank) [inline Looms.PoS/Looms.PoS.Application/Features/Discount/Queries/GetDiscounts/GetDiscountsQuery.cs:6]: new line after namespace

**Review by titasstank — DISMISSED** [review]

File: Looms.PoS/Looms.PoS.Application/Features/Discount/Queries/GetDiscounts/GetDiscountsQuery.cs (hunk 15 above)
Comment (aiste-i) [reply Looms.PoS/Looms.PoS.Application/Features/Discount/Queries/GetDiscounts/GetDiscountsQuery.cs:6]: New line added after namespace

**Review by titasstank — APPROVED** [review]

**Review by epuluikis — APPROVED** [review]

**Pull request merged**